augur/tasks/util/collection_util.py

```diff
@@ -425,3 +433,155 @@ def start_block_of_repos(logger,session,condition,limit,phases,repos_type,hook="
     routine.start_data_collection()
 
     return len(repo_git_identifiers)
+
+"""
+    Generalized function for starting a phase of tasks for a given collection hook with options to add restrictive conditions
+"""
+def start_repos_by_user(session, max_repo,phase_list, days_until_collect_again = 1, hook="core",new_status=CollectionState.PENDING.value,additional_conditions=None):
+
+    #getattr(CollectionStatus,f"{hook}_status" ) represents the status of the given hook
+    #Get the count of repos that are currently running this collection hook
+    desired_status = f"{hook}_status"
+    active_repo_count = len(session.query(CollectionStatus).filter(getattr(CollectionStatus,desired_status ) == CollectionState.COLLECTING.value).all())
+
+    #Will always disallow errored repos and repos that are already collecting
+    not_erroed = getattr(CollectionStatus,desired_status ) != str(CollectionState.ERROR.value)
+    not_collecting = getattr(CollectionStatus,desired_status ) != str(CollectionState.COLLECTING.value)
+
+    #The maximum amount of repos to schedule is affected by the existing repos running tasks
+    limit = max_repo-active_repo_count
+
+    #Split all users that have new repos into four lists and randomize order
+    query = s.sql.text(f"""
+        SELECT  
+        user_id
+        FROM augur_operations.user_groups 
+        JOIN augur_operations.user_repos ON augur_operations.user_groups.group_id = augur_operations.user_repos.group_id
+        JOIN augur_data.repo ON augur_operations.user_repos.repo_id = augur_data.repo.repo_id
+        JOIN augur_operations.collection_status ON augur_operations.user_repos.repo_id = augur_operations.collection_status.repo_id
+        WHERE {desired_status}='{str(new_status)}'
+        GROUP BY user_id
+    """)
+
+    user_list = session.execute_sql(query).fetchall()
+    random.shuffle(user_list)
+
+    #Extract the user id from the randomized list and split into four chunks
+    split_user_list = split_list_into_chunks([row[0] for row in user_list], 4)
+
+    session.logger.info(f"User_list: {split_user_list}")
+
+    #Iterate through each fourth of the users fetched
+    for quarter_list in split_user_list:
+        if limit <= 0:
+            return
+
+        #Query a set of valid repositories sorted by weight, also making sure that the repos are new
```

Comment:
So is this *only* going to run new repos?

Reply:
It will always run new repos first if they exist, and then if we didn't hit the max repo count when scheduling new repos we schedule more old repos until we hit the max repo count or if there aren't enough old repos (repos collected more than a day ago) we just schedule as many as we can